### ISSUE TITLE
Odd mismatch b/n 2D bboxes and images

### ISSUE THREAD

**Issue by Divadi** [once-for-auto-driving/once_devkit#9]
I have visualized some of the 2D bounding boxes onto the undistorted images, and I got some odd results:
![image](https://user-images.githubusercontent.com/17109672/126398468-08335397-5e55-4e56-b50d-087c35abb554.png)
The truck annotation on the left seems to be a bit mismatched, and zooming into the cyclists annotations, there doesn't seem to be anything there:
![image](https://user-images.githubusercontent.com/17109672/126398613-c9207e3d-f043-40bc-8e75-f812cc6daeda.png)

Some other curiosities:
There seems to be some annotations for objects completely occluded (left):
![image](https://user-images.githubusercontent.com/17109672/126398753-2ee86138-6f1b-4a25-971c-65f9263c8f16.png)
Offset boxes:
![image](https://user-images.githubusercontent.com/17109672/126398900-f8d26ea1-4112-4e39-b2f9-342def14d170.png)
Odd:
![image](https://user-images.githubusercontent.com/17109672/126426691-17d8d8f7-ce91-4f7b-b82b-36c26f4bf970.png)

Did I perhaps make a transformation error/is there anything I can do to remedy this issue?
Also, would it be possible to have access to the code generating the 2D boxes?

**Comment by Divadi** [comment]
Okay, I changed `once.py`'s undistort image function:
```
    def undistort_image(self, seq_id, frame_id):
        img_list = []
        split_name = self._find_split_name(seq_id)
        frame_info = getattr(self, '{}_info'.format(split_name))[seq_id][frame_id]
        for cam_name in self.__class__.camera_names:
            img_buf = self.load_image(seq_id, frame_id, cam_name)
            cam_calib = frame_info['calib'][cam_name]
            h, w = img_buf.shape[:2]
            # Previous 
#             cv2.getOptimalNewCameraMatrix(cam_calib['cam_intrinsic'],
#                                           cam_calib['distortion'],
#                                           (w, h), alpha=0.0, newImgSize=(w, h))))
#             img_list.append(cv2.undistort(img_buf, cam_calib['cam_intrinsic'],
#                                           cam_calib['distortion'],
#                                           newCameraMatrix=cam_calib['cam_intrinsic']))
            # New
            cam_calib['new_cam_intrinsic'], _ = cv2.getOptimalNewCameraMatrix(cam_calib['cam_intrinsic'],
                                          cam_calib['distortion'],
                                          (w, h), alpha=0.0, newImgSize=(w, h))
            img_list.append(cv2.undistort(img_buf, cam_calib['cam_intrinsic'],
                                          cam_calib['distortion'],
                                          newCameraMatrix=cam_calib['new_cam_intrinsic']))
        return img_list
```
To use the new cam intrinsic for undistorting the image.
Doing so, I was able to get seemingly correct 2D boxes:
![image](https://user-images.githubusercontent.com/17109672/126420690-0f801d2a-7b81-44a2-a2a9-48c1fbd86bf2.png)
![image](https://user-images.githubusercontent.com/17109672/126420706-1a2fc7ce-8b49-4ffc-8bb4-a07e3bc12672.png)
![image](https://user-images.githubusercontent.com/17109672/126420730-56afd405-4dbf-4ba4-8c5f-2b452808efd9.png)

To check, I also wanted to project the 3D boxes myself. I projected using the new_cam_intrinsic matrix and they seem quite close:
Colored boxes are 3D, black boxes are 2D:
![image](https://user-images.githubusercontent.com/17109672/126420954-036f1206-a1e1-42fd-9230-56c685cfb38d.png)
However, some boxes don't line up:
![image](https://user-images.githubusercontent.com/17109672/126431061-863f49f0-a71d-40a5-85df-1e67d429491d.png)![image](https://user-images.githubusercontent.com/17109672/126431078-5cf34403-d431-40d5-9d7e-1c1be016fc3c.png)

And for cam07 images, my projected boxes seem to be "lower" for all sequences:
![image](https://user-images.githubusercontent.com/17109672/126421019-24017b24-46d2-4ddb-9287-52723175b090.png)
![image](https://user-images.githubusercontent.com/17109672/126431170-b36b618a-e504-4e26-83aa-fc220131c7ec.png)

I am hoping to get some guidance on this

**Comment by once-for-auto-driving** [comment]
Thanks for your attention to our dataset!

Offset boxes: We tested the image undistortion and projection using OpenCV (4.1.1) and didn't find such an issue, so we guess it is a problem with different OpenCV versions. Could you please provide the environment you use?

Occluded boxes: Our Lidar sensor is installed at a higher position compared to the cameras, so some objects that the lidar sensor can "see" may be occluded on the images. We annotated 3D boxes in point clouds captured by the lidar sensor.

Missing/Odd boxes: There are still some annotation errors even after two rounds of humans' checks. We plan to give another check to the annotations asap. Considering we are holding a workshop and a challenge, we may update the annotations later in this/next year.

**Comment by nmzfrank** [comment]
Please refer to new commit https://github.com/once-for-auto-driving/once_devkit/blob/b659f4d66b5a3dceeaae9b8c66ae441e9379a8a3/once.py#L188. Misalign on cam07 is due to wrong distortion process in your code.

**Comment by Divadi** [comment]
Thank you for your prompt response!

> Thanks for your attention to our dataset!
> 
> Offset boxes: We tested the image undistortion and projection using OpenCV (4.1.1) and didn't find such an issue, so we guess it is a problem with different OpenCV versions. Could you please provide the environment you use?
> 
> Occluded boxes: Our Lidar sensor is installed at a higher position compared to the cameras, so some objects that the lidar sensor can "see" may be occluded on the images. We annotated 3D boxes in point clouds captured by the lidar sensor.
> 
> Missing/Odd boxes: There are still some annotation errors even after two rounds of humans' checks. We plan to give another check to the annotations asap. Considering we are holding a workshop and a challenge, we may update the annotations later in this/next year.

Offset boxes: I had been using 4.2, I changed to 4.1.1 and there was no difference
Occluded boxes: I understand! So they are visible to the lidar but not the camera

------------------------
> Please refer to new commit
> 
> https://github.com/once-for-auto-driving/once_devkit/blob/b659f4d66b5a3dceeaae9b8c66ae441e9379a8a3/once.py#L188
> 
> . Misalign on cam07 is due to wrong distortion process in your code.

Running `once.py` as is, I had to change a few things:
I only have train & val downloaded, so only looped over those here:
https://github.com/once-for-auto-driving/once_devkit/blob/b659f4d66b5a3dceeaae9b8c66ae441e9379a8a3/once.py#L100

It seems like in the json, frames with annotations have frame_id as a string value, and frames without annotations have it as an int value, which causes isuses here
https://github.com/once-for-auto-driving/once_devkit/blob/b659f4d66b5a3dceeaae9b8c66ae441e9379a8a3/once.py#L112
and when when sorting here:
https://github.com/once-for-auto-driving/once_devkit/blob/b659f4d66b5a3dceeaae9b8c66ae441e9379a8a3/once.py#L125
So, I just changed line 112 to 
```
frame_list.append(str(frame_anno['frame_id']))
```


Then I changed this as I think rectangle takes a start point & end point
https://github.com/once-for-auto-driving/once_devkit/blob/b659f4d66b5a3dceeaae9b8c66ae441e9379a8a3/once.py#L203
to
```
cv2.rectangle(img_buf, tuple(box2d[:2]), tuple(box2d[2:]), (255, 0, 0), 2)
```

------------------------
With this, for seq 000076, frame 1616343527200, cam03:
![image](https://user-images.githubusercontent.com/17109672/126566684-bb07df75-ba13-4f73-9784-0d4114f92128.png)

As I mentioned above, **it seems like in `undistort_image`, the call to `cv2.getOptimalNewCameraMatrix` seems to not be doing anything:**
https://github.com/once-for-auto-driving/once_devkit/blob/b659f4d66b5a3dceeaae9b8c66ae441e9379a8a3/once.py#L144-L157
I thought it should return a new camera matrix & roi, and the new camera matrix should be used for undistort like this (with roi ignored since alpha is 0.0):
```
def undistort_image(self, seq_id, frame_id):
        img_list = []
        split_name = self._find_split_name(seq_id)
        frame_info = getattr(self, '{}_info'.format(split_name))[seq_id][frame_id]
        for cam_name in self.__class__.camera_names:
            img_buf = self.load_image(seq_id, frame_id, cam_name)
            cam_calib = frame_info['calib'][cam_name]
            h, w = img_buf.shape[:2]
            
            cam_calib['new_cam_intrinsic'], _ = cv2.getOptimalNewCameraMatrix(cam_calib['cam_intrinsic'],
                                          cam_calib['distortion'],
                                          (w, h), alpha=0.0, newImgSize=(w, h))

            img_list.append(cv2.undistort(img_buf, cam_calib['cam_intrinsic'],
                                          cam_calib['distortion'],
                                          newCameraMatrix=cam_calib['new_cam_intrinsic']))
        return img_list
```
Doing so yields 
![image](https://user-images.githubusercontent.com/17109672/126567149-66dea2e7-7a22-46d2-bd01-ff0c8aaecc35.png)

Please let me know if my corrections are wrong - even with these changes, I have the issue of misalignment b/n 3D & 2D boxes for some frames

**Comment by nmzfrank** [comment]
Hello @Divadi , thanks for your suggestion. We find out that the 2D projection is different from 3D lidar projection onto images and update a new resort image method to solve the problem as well as the 'str' and 'coordinate' bug you mensioned before.

As we mentioned, the 2D bounding boxes is the projection from 3D bounding box and is generated by annotation system automatically with no human check. As a result some 2D boxes which is 'behind' the camera may be reserved, causing some odd bounding box projects. So currently it is not recommended to use these noisy 2D annotations for supervision signals.

**Comment by Divadi** [comment]
Thank you for these changes, and thank you for your patience. I have one last error, and this is regarding cam07. (Again, sorry for this long comment, but I had to include visualizations)

I understand that 2D bounding boxes are generated by projecting 3D boxes and that there may be errors. However, the 3D projected boxes generally do include the object, albeit not tightly, and I am hoping to use this for additional supervision. I am hoping to project the 3D boxes and resolve some of the incomplete annotations myself.
At the very least, I am trying to setup projection from 3D to 2D for lidar.

Here are some cam07 outputs from the given `project_lidar_to_image` function:
`'000092', '1616442487300'`
![image](https://user-images.githubusercontent.com/17109672/126848046-a3e2bf52-e523-489c-a5a8-32a03b4eaf5e.png)
`'000076', '1616343527200'`
![image](https://user-images.githubusercontent.com/17109672/126848150-3dc3e58c-35d4-4fa5-9c81-42fb8aa1015e.png)

I notice that there seems to be misalignment (I observe no such issue for other camera frames).
So, I made a small adjustment to only take points _within_ a 3D bbox annotation:
```
def project_lidar_to_image(self, seq_id, frame_id, only_pts_in_box=True):
        points = self.load_point_cloud(seq_id, frame_id)
        
        if only_pts_in_box:
            import torch
            from pcdet.ops.roiaware_pool3d.roiaware_pool3d_utils import points_in_boxes_cpu
            annos = self.get_frame_anno(seq_id, frame_id)
            mask = points_in_boxes_cpu(points[:, :3], torch.tensor(annos['boxes_3d']))
            mask = torch.any(mask, dim=0) # point is in any box
            points = points[mask]

       ... rest is same...
```
Doing so, the misalignment is clear:
![image](https://user-images.githubusercontent.com/17109672/126848329-41ce9120-3a54-4e83-8ce9-01850197857f.png)
![image](https://user-images.githubusercontent.com/17109672/126848346-bb352b7a-d424-49af-b1b3-8fa2d03e87c5.png)

Some other examples:
`'000076', '1616343844199'`
![image](https://user-images.githubusercontent.com/17109672/126848394-14f7ab67-d1c6-405d-8c42-2d19df2fc58d.png)
`'000092', '1616442892300'`
![image](https://user-images.githubusercontent.com/17109672/126848484-ac4041c6-0074-497a-aae4-66a0e44515b5.png)
`'000092', '1616443070800'`
![image](https://user-images.githubusercontent.com/17109672/126848551-716560e2-c230-4c0c-b53f-830f62fe94e8.png)
![image](https://user-images.githubusercontent.com/17109672/126848556-fedefa65-fc9d-4bed-93a9-75395810e432.png)
`'000092', '1616442955301'`
![image](https://user-images.githubusercontent.com/17109672/126848587-91e9b4ba-649e-44cb-990a-24f2435796e5.png)

For other cameras, the alignment is extremely precise:
![image](https://user-images.githubusercontent.com/17109672/126848603-78ee5e83-086e-4308-829f-d88b53723f35.png)
![image](https://user-images.githubusercontent.com/17109672/126848609-dec661b0-63cd-4c8e-ab1f-298afea8937b.png)
![image](https://user-images.githubusercontent.com/17109672/126848627-81801143-9cff-4dca-83ae-bb9f8ca7ae97.png)
So I feel it is an issue with something about cam07.

However, I find that **the cam07 2D bbox labels are very accurate**, while my projected 3D boxes instead match the points-in-boxes (so mis-matched with 2D annotations). 
So, the methodology used to generate the 2D box annotations were probably correct, but is perhaps not what is used in `once.py`?
(please see last two images in my second comment above, and here are more samples):
black boxes are 2D annos in json, colored 3D boxes are boxes I projected, and only points in some 3d box are shown:
![image](https://user-images.githubusercontent.com/17109672/126848793-4b030367-b87b-41fc-9720-a3de4c656aa8.png)
![image](https://user-images.githubusercontent.com/17109672/126848899-6ac0988f-f30e-49e5-8c2e-0a67386693ee.png)

And to show that my script does not have bugs, I use the exact same script for other cameras:
![image](https://user-images.githubusercontent.com/17109672/126848994-f2c80880-8d01-4c90-8111-017ce2a3be53.png)
![image](https://user-images.githubusercontent.com/17109672/126849002-03b87412-6199-4e23-9ebc-40ca41dfbfd4.png)

Please let me know if I made an error/if there is something I can do from my side to rectify this cam07 mismatch.
I apologize for the long post

**Comment by nmzfrank** [comment]
Hello @Divadi , thanks for your detailed analyze and feel free for the long post as long as you feel it necessary to provide the images. 
Actually the misalignment of cam07 comes from the movement of the vehicle. While Lidar rotates one round to collect one frame, images are token by cameras. Though we try to keep sync between camera and lidar, there is some time gap between the moment images are shot and lidar points are returned, which means some image are token 'after' lidar points. This is perticular obvious when the speed of vehicle is high or the vehicle is turning. Take '000092', '1616443070800' for example, actually if you check the other cameras of the same frame you can find out that the misalignment happens too.


As for the 3D bounding box and 2D box mismatch, currently we provide two undistort method for 3D and 2D seperatly. You should use undistort_image for 3D bounding box project and undistort_image_v2 for 2D box. So I think 3D and 2D box can not draw in the same image.

**Comment by Divadi** [comment]
Really, thank you so much for your patience and efforts.

I see; that does make sense.
Visualizing that frame:
cam7,8,9:
![image](https://user-images.githubusercontent.com/17109672/126937919-48d0b72b-c54f-4740-a2c6-f41bac23022c.png)
![image](https://user-images.githubusercontent.com/17109672/126937932-36e9675d-84f5-4a16-9202-9a977347d397.png)
![image](https://user-images.githubusercontent.com/17109672/126937944-f4837975-50ae-49fd-b096-d5f97b502f44.png)
8 seems alright, but I can see that in 9, the left-most pedestrian has some offset. Although it is a bit weird that cam07 seems to most consistently have this issue.

Lastly, beyond this issue, the 2D boxes were generated from 3D box projection without human intervention. So, there must be _some_ way 3D can accurately be projected onto cam07, right? Because I see that 2D boxes on cam07 are very good (no offset issue like there is for 3D), examples shown way above. I have tried 3D projection for both undistort_image and undistort_image_v2, as well as the old and new camera intrinsics (all four combinations of this), but the 3D and 2D do not line up for any of them. How were the high-quality cam07 2D annotations generated?

**Comment by nmzfrank** [comment]
Hello @Divadi , firstly,  I'd like to thank you for your patient and careful check on ONCE dataset. 
It takes some time to re-check the rawdata and calibration information.  We find out that calibration info to the part of the data is wrong, which causes the misalignment of lidar projection for cam07. We update all the six info files (which contains all the calibration data) to baidu Yunpan and google drive. You can download the updated files through [ONCE dataset](https://once-for-auto-driving.github.io/download.html#downloads). 
We also correct the projection function for both 2D and 3D annotations in https://github.com/once-for-auto-driving/once_devkit/blob/c084b5d49a029e34570d4ddb7b7011052cb955ee/once.py#L212, which uses the new camera intrinsic matrix and now 2D and 3D boxes matches.
